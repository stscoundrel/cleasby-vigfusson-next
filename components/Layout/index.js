// Components.
import Head from 'components/Head'
import Footer from 'components/Footer'
import Breadcrumbs from 'components/Breadcrumbs'
import Navigation from 'components/Navigation'

export default function Layout({
  type, content, children, letters, noSearch = false,
}) {
  return (
    <>
      <Head type={type} content={content}></Head>
      <main>
        <Navigation letters={letters} noSearch={noSearch}/>
        <Breadcrumbs type={type} content={content} />
        {children}
      </main>
      <Footer letters={letters}/>
    </>
  )
}
