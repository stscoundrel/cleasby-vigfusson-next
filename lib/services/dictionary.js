const { getDictionary } = require('cleasby-vigfusson-dictionary');
const slugify = require('slugify');

const addSlugs = (words) => {
  const existingSlugs = {};

  const formattedWords = words.map((word) => {
    let slug = slugify(word.word).toLowerCase();

    if (existingSlugs[slug]) {
      // Double slug, make unique.
      existingSlugs[slug] += 1;
      slug = `${slug}-${existingSlugs[slug]}`;
    } else {
      existingSlugs[slug] = 1;
    }

    return {
      ...word,
      slug,
    };
  });

  return formattedWords;
};

export const getAllWords = () => {
  const words = getDictionary();

  /**
   * Add URL safe slugs.
   */
  const formattedWords = addSlugs(words);

  return formattedWords;
};

export default {
  getAllWords,
};
