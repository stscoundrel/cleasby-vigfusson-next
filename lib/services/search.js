export const searchDictionary = (search, dictionary) => {
  const results = dictionary.filter((entry) => {
    let matchesSearch = false

    if (entry.word.includes(search)) {
      matchesSearch = true
    }

    if (entry.slug.includes(search)) {
      matchesSearch = true
    }

    entry.definitions.forEach((definition) => {
      if (definition.includes(search)) {
        matchesSearch = true
      }
    })

    return matchesSearch
  })

  return results
}

export default {
  searchDictionary,
}
