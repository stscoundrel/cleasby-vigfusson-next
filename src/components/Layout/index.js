// Components.
import Head from 'components/Head'
import Footer from 'components/Footer'
import Breadcrumbs from 'components/Breadcrumbs'
import Navigation from 'components/Navigation'
import BackToTop from 'components/BackToTop'

export default function Layout({
  type, content, children, letters, letter = false, noSearch = false,
}) {
  return (
    <>
      <Head type={type} content={content} letter={letter}></Head>
      <header>
        <Navigation letters={letters} noSearch={noSearch}/>
        <Breadcrumbs type={type} content={content} />
      </header>
      <main className="container">
        {children}
        <BackToTop />
      </main>
      <Footer letters={letters}/>
    </>
  )
}
