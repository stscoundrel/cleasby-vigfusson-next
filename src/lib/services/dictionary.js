const { getDictionary } = require('cleasby-vigfusson-dictionary')
const { VALID_AS_FIRST } = require('old-norse-alphabet')
const { slugifyWord, slugifyLetter } = require('../utils/slugs')

let cachedDictionary = null

const addSlugs = (words) => {
  const existingSlugs = {}

  const formattedWords = words.map((word) => {
    let slug = slugifyWord(word.word).toLowerCase()

    if (slug.charAt(0) === '-') {
      slug = slug.substring(1)
    }

    if (existingSlugs[slug]) {
      // Double slug, make unique.
      existingSlugs[slug] += 1
      slug = `${slug}-${existingSlugs[slug]}`
    } else {
      existingSlugs[slug] = 1
    }

    return {
      ...word,
      slug,
    }
  })

  return formattedWords
}

export const getAllWords = () => {
  if (cachedDictionary) return cachedDictionary

  const words = getDictionary()

  /**
   * Add URL safe slugs.
   */
  const formattedWords = addSlugs(words)

  cachedDictionary = formattedWords

  return formattedWords
}

export const getByLetter = (letter) => {
  const words = getAllWords()
  const byLetter = words.filter((entry) => (
    entry.word.charAt(0).toLowerCase() === letter.toLowerCase()))

  return byLetter
}

export const getWord = (word) => getAllWords().filter((entry) => entry.slug === word)[0]

export const getAlphabet = () => {
  const letters = [...VALID_AS_FIRST.filter((letter) => letter !== 'ǫ' && letter !== 'ø'), 'ö']

  const formattedLetters = letters.map((letter) => ({
    letter,
    slug: slugifyLetter(letter),
  }))

  return formattedLetters
}

export default {
  getAllWords,
  getByLetter,
  getWord,
  getAlphabet,
}
