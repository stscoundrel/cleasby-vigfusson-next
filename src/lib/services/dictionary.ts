import { getDictionary, DictionaryEntry as RawDictionaryEntry } from 'cleasby-vigfusson-dictionary'
import { VALID_AS_FIRST } from 'old-norse-alphabet'
import { oldNorseSort } from 'old-norse-alphabet-sort'
import { slugifyWord, slugifyLetter } from '../utils/slugs'

export interface DictionaryEntry extends RawDictionaryEntry {
  slug: string,
}

export interface DictionaryEntryDTO {
  word: string,
  slug: string
}

export interface AlphabetLetter {
  letter: string,
  slug: string
}

let cachedDictionary: DictionaryEntry[] | null = null

const addSlugs = (words: RawDictionaryEntry[]): DictionaryEntry[] => {
  const existingSlugs = {}

  const formattedWords = words.map((word) => {
    let slug = slugifyWord(word.word).toLowerCase()

    if (slug.charAt(0) === '-') {
      slug = slug.substring(1)
    }

    if (existingSlugs[slug]) {
      // Double slug, make unique.
      existingSlugs[slug] += 1
      slug = `${slug}-${existingSlugs[slug]}`
    } else {
      existingSlugs[slug] = 1
    }

    return {
      ...word,
      slug,
    }
  })

  return formattedWords
}

export const getAllWords = (): DictionaryEntry[] => {
  if (cachedDictionary) return cachedDictionary

  const words = getDictionary()

  /**
   * Add URL safe slugs.
   */
  const formattedWords = addSlugs(words)

  /**
   * Ensure headwords are lowercased.
   * Original source has some in all caps.
   * That is true to the layout of the book,
   * but makes little sense in website.
   */
  const dictionary = formattedWords.map((entry) => ({ ...entry, word: entry.word.toLowerCase() }))

  cachedDictionary = dictionary

  return dictionary
}

export const getByLetter = (letter: string): DictionaryEntryDTO[] => {
  const words = getAllWords()
  const byLetter = words
    .filter((entry) => (
      entry.word.charAt(0).toLowerCase() === letter.toLowerCase()))
    .map((entry) => {
      // Simpler DTO dictionary entry.
      const { word, slug } = entry
      return { word, slug }
    })
    .sort((a, b) => oldNorseSort(a.word, b.word))

  return byLetter
}

export const getWord = (slug: string): DictionaryEntry => (
  getAllWords().filter((entry) => entry.slug === slug)[0]
)

export const getRandomEntries = (): DictionaryEntry[] => (
  // Return entries fit to be randomized "teasers"
  // Therefore, content should be short, but not too short.
  getAllWords()
    .sort(() => Math.random() - 0.5)
    .filter((entry) => entry.definitions[0].length < 50 && entry.definitions[0].length > 15)
    .slice(0, 36)
    .sort((a, b) => oldNorseSort(a.word, b.word))
)

export const getSimilarWords = (entry: DictionaryEntry) : DictionaryEntry[] => getAllWords()
  .filter((dEntry) => dEntry.word.toLocaleLowerCase() === entry.word.toLocaleLowerCase()
  && dEntry.slug !== entry.slug)

export const getAlphabet = (): AlphabetLetter[] => {
  const letters = [...VALID_AS_FIRST.filter((letter) => letter !== 'ǫ' && letter !== 'ø'), 'ö']

  const formattedLetters = letters.map((letter) => ({
    letter,
    slug: slugifyLetter(letter),
  }))

  return formattedLetters
}

export default {
  getAllWords,
  getByLetter,
  getWord,
  getRandomEntries,
  getAlphabet,
}
