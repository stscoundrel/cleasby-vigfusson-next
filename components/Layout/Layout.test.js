import ReactDOM from 'react-dom'
import Layout from 'components/Layout'
import renderer from 'react-test-renderer'
import { getByLetter, getWord, getAlphabet } from 'lib/services/dictionary'

describe('Layout component', () => {
  describe('Letter layout', () => {
    const aWords = getByLetter('a').slice(0, 10)

    test('Does not crash', () => {
      const div = document.createElement('div')
      ReactDOM.render(<Layout content={aWords} type="letter" letters={getAlphabet()} />, div)
      ReactDOM.unmountComponentAtNode(div)
    })

    test('Matches snapshot', () => {
      const tree = renderer.create(<Layout content={aWords} type="letter" letters={getAlphabet()} />).toJSON()
      expect(tree).toMatchSnapshot()
    })
  })

  describe('Word layout', () => {
    const word = getWord('skilja')

    test('Does not crash', () => {
      const div = document.createElement('div')
      ReactDOM.render(<Layout content={word} type="word" letters={getAlphabet()} />, div)
      ReactDOM.unmountComponentAtNode(div)
    })

    test('Matches snapshot', () => {
      const tree = renderer.create(<Layout content={word} type="word" letters={getAlphabet()} />).toJSON()
      expect(tree).toMatchSnapshot()
    })
  })
})
