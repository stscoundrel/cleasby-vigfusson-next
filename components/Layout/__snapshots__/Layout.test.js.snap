// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Layout component Letter layout Matches snapshot 1`] = `
Array [
  <main>
    <nav>
      <ul
        className="list"
      >
        <li
          className="listItem"
        >
          <a
            href="/letter/a"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            a
          </a>
        </li>
        <li
          className="listItem"
        >
          <a
            href="/letter/a2"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            á
          </a>
        </li>
        <li
          className="listItem"
        >
          <a
            href="/letter/b"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            b
          </a>
        </li>
        <li
          className="listItem"
        >
          <a
            href="/letter/d"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            d
          </a>
        </li>
        <li
          className="listItem"
        >
          <a
            href="/letter/e"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            e
          </a>
        </li>
        <li
          className="listItem"
        >
          <a
            href="/letter/e2"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            é
          </a>
        </li>
        <li
          className="listItem"
        >
          <a
            href="/letter/f"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            f
          </a>
        </li>
        <li
          className="listItem"
        >
          <a
            href="/letter/g"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            g
          </a>
        </li>
        <li
          className="listItem"
        >
          <a
            href="/letter/h"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            h
          </a>
        </li>
        <li
          className="listItem"
        >
          <a
            href="/letter/i"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            i
          </a>
        </li>
        <li
          className="listItem"
        >
          <a
            href="/letter/i2"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            í
          </a>
        </li>
        <li
          className="listItem"
        >
          <a
            href="/letter/j"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            j
          </a>
        </li>
        <li
          className="listItem"
        >
          <a
            href="/letter/k"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            k
          </a>
        </li>
        <li
          className="listItem"
        >
          <a
            href="/letter/l"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            l
          </a>
        </li>
        <li
          className="listItem"
        >
          <a
            href="/letter/m"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            m
          </a>
        </li>
        <li
          className="listItem"
        >
          <a
            href="/letter/n"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            n
          </a>
        </li>
        <li
          className="listItem"
        >
          <a
            href="/letter/o"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            o
          </a>
        </li>
        <li
          className="listItem"
        >
          <a
            href="/letter/o2"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            ó
          </a>
        </li>
        <li
          className="listItem"
        >
          <a
            href="/letter/p"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            p
          </a>
        </li>
        <li
          className="listItem"
        >
          <a
            href="/letter/r"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            r
          </a>
        </li>
        <li
          className="listItem"
        >
          <a
            href="/letter/s"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            s
          </a>
        </li>
        <li
          className="listItem"
        >
          <a
            href="/letter/t"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            t
          </a>
        </li>
        <li
          className="listItem"
        >
          <a
            href="/letter/u"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            u
          </a>
        </li>
        <li
          className="listItem"
        >
          <a
            href="/letter/u2"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            ú
          </a>
        </li>
        <li
          className="listItem"
        >
          <a
            href="/letter/v"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            v
          </a>
        </li>
        <li
          className="listItem"
        >
          <a
            href="/letter/y"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            y
          </a>
        </li>
        <li
          className="listItem"
        >
          <a
            href="/letter/y2"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            ý
          </a>
        </li>
        <li
          className="listItem"
        >
          <a
            href="/letter/th"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            þ
          </a>
        </li>
        <li
          className="listItem"
        >
          <a
            href="/letter/ae"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            æ
          </a>
        </li>
        <li
          className="listItem"
        >
          <a
            href="/letter/oe3"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            œ
          </a>
        </li>
        <li
          className="listItem"
        >
          <a
            href="/letter/oe"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            ö
          </a>
        </li>
      </ul>
      <form
        onSubmit={[Function]}
      >
        <input
          name="search"
          onChange={[Function]}
          type="text"
          value=""
        />
        <button
          type="submit"
        >
          Search
        </button>
      </form>
    </nav>
    <div>
      <a
        href="/"
        onClick={[Function]}
        onMouseEnter={[Function]}
      >
        Cleasby & Vigfusson Dictionary
      </a>
      <a
        href="/letter/a"
        onClick={[Function]}
        onMouseEnter={[Function]}
      >
        Letter A
      </a>
      <script
        dangerouslySetInnerHTML={
          Object {
            "__html": "{\\"@context\\":\\"https://schema.org/\\",\\"@type\\":\\"BreadcrumbList\\",\\"itemListElement\\":[{\\"@type\\":\\"ListItem\\",\\"position\\":1,\\"name\\":\\"Cleasby & Vigfusson Dictionary\\",\\"item\\":\\"undefined/\\"},{\\"@type\\":\\"ListItem\\",\\"position\\":2,\\"name\\":\\"Letter A\\",\\"item\\":\\"undefined/letter/a\\"}]}",
          }
        }
        type="application/ld+json"
      />
    </div>
  </main>,
  <footer>
    <hr />
    <section>
      <h2>
        About
      </h2>
      <p>
        Based on Cleasby & Vigfusson Dictionary Old Norse dictionary.
      </p>
      <p>
        <em>
          Icelandic-English
        </em>
         dictionary was started by Richard Cleasby and finished by Gudbrand Vigfusson.
      </p>
      <p>
        It was published in 1874, which leads to there being many public domain versions of the book available.
      </p>
    </section>
    <nav>
      <p>
        Dictionary project
      </p>
      <ul>
        <li>
          <a
            href="https://github.com/stscoundrel/cleasby-vigfusson-next"
            rel="noopener noreferrer"
            target="_blank"
          >
            Source code
          </a>
        </li>
        <li>
          <a
            href="https://github.com/stscoundrel/cleasby-vigfusson-dictionary"
            rel="noopener noreferrer"
            target="_blank"
          >
            Data source
          </a>
        </li>
      </ul>
    </nav>
    <nav>
      <p>
        Quick links
      </p>
      <ul>
        <li>
          <a
            href="/letter/a"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            a
          </a>
        </li>
        <li>
          <a
            href="/letter/a2"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            á
          </a>
        </li>
        <li>
          <a
            href="/letter/b"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            b
          </a>
        </li>
        <li>
          <a
            href="/letter/d"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            d
          </a>
        </li>
        <li>
          <a
            href="/letter/e"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            e
          </a>
        </li>
        <li>
          <a
            href="/letter/e2"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            é
          </a>
        </li>
        <li>
          <a
            href="/letter/f"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            f
          </a>
        </li>
        <li>
          <a
            href="/letter/g"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            g
          </a>
        </li>
        <li>
          <a
            href="/letter/h"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            h
          </a>
        </li>
        <li>
          <a
            href="/letter/i"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            i
          </a>
        </li>
        <li>
          <a
            href="/letter/i2"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            í
          </a>
        </li>
        <li>
          <a
            href="/letter/j"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            j
          </a>
        </li>
        <li>
          <a
            href="/letter/k"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            k
          </a>
        </li>
        <li>
          <a
            href="/letter/l"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            l
          </a>
        </li>
        <li>
          <a
            href="/letter/m"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            m
          </a>
        </li>
        <li>
          <a
            href="/letter/n"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            n
          </a>
        </li>
        <li>
          <a
            href="/letter/o"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            o
          </a>
        </li>
        <li>
          <a
            href="/letter/o2"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            ó
          </a>
        </li>
        <li>
          <a
            href="/letter/p"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            p
          </a>
        </li>
        <li>
          <a
            href="/letter/r"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            r
          </a>
        </li>
        <li>
          <a
            href="/letter/s"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            s
          </a>
        </li>
        <li>
          <a
            href="/letter/t"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            t
          </a>
        </li>
        <li>
          <a
            href="/letter/u"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            u
          </a>
        </li>
        <li>
          <a
            href="/letter/u2"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            ú
          </a>
        </li>
        <li>
          <a
            href="/letter/v"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            v
          </a>
        </li>
        <li>
          <a
            href="/letter/y"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            y
          </a>
        </li>
        <li>
          <a
            href="/letter/y2"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            ý
          </a>
        </li>
        <li>
          <a
            href="/letter/th"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            þ
          </a>
        </li>
        <li>
          <a
            href="/letter/ae"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            æ
          </a>
        </li>
        <li>
          <a
            href="/letter/oe3"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            œ
          </a>
        </li>
        <li>
          <a
            href="/letter/oe"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            ö
          </a>
        </li>
      </ul>
    </nav>
    <small>
      Copyright © 2020 Sampo Silvennoinen / StScoundrel
    </small>
  </footer>,
]
`;

exports[`Layout component Word layout Matches snapshot 1`] = `
Array [
  <main>
    <nav>
      <ul
        className="list"
      >
        <li
          className="listItem"
        >
          <a
            href="/letter/a"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            a
          </a>
        </li>
        <li
          className="listItem"
        >
          <a
            href="/letter/a2"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            á
          </a>
        </li>
        <li
          className="listItem"
        >
          <a
            href="/letter/b"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            b
          </a>
        </li>
        <li
          className="listItem"
        >
          <a
            href="/letter/d"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            d
          </a>
        </li>
        <li
          className="listItem"
        >
          <a
            href="/letter/e"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            e
          </a>
        </li>
        <li
          className="listItem"
        >
          <a
            href="/letter/e2"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            é
          </a>
        </li>
        <li
          className="listItem"
        >
          <a
            href="/letter/f"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            f
          </a>
        </li>
        <li
          className="listItem"
        >
          <a
            href="/letter/g"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            g
          </a>
        </li>
        <li
          className="listItem"
        >
          <a
            href="/letter/h"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            h
          </a>
        </li>
        <li
          className="listItem"
        >
          <a
            href="/letter/i"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            i
          </a>
        </li>
        <li
          className="listItem"
        >
          <a
            href="/letter/i2"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            í
          </a>
        </li>
        <li
          className="listItem"
        >
          <a
            href="/letter/j"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            j
          </a>
        </li>
        <li
          className="listItem"
        >
          <a
            href="/letter/k"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            k
          </a>
        </li>
        <li
          className="listItem"
        >
          <a
            href="/letter/l"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            l
          </a>
        </li>
        <li
          className="listItem"
        >
          <a
            href="/letter/m"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            m
          </a>
        </li>
        <li
          className="listItem"
        >
          <a
            href="/letter/n"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            n
          </a>
        </li>
        <li
          className="listItem"
        >
          <a
            href="/letter/o"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            o
          </a>
        </li>
        <li
          className="listItem"
        >
          <a
            href="/letter/o2"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            ó
          </a>
        </li>
        <li
          className="listItem"
        >
          <a
            href="/letter/p"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            p
          </a>
        </li>
        <li
          className="listItem"
        >
          <a
            href="/letter/r"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            r
          </a>
        </li>
        <li
          className="listItem"
        >
          <a
            href="/letter/s"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            s
          </a>
        </li>
        <li
          className="listItem"
        >
          <a
            href="/letter/t"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            t
          </a>
        </li>
        <li
          className="listItem"
        >
          <a
            href="/letter/u"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            u
          </a>
        </li>
        <li
          className="listItem"
        >
          <a
            href="/letter/u2"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            ú
          </a>
        </li>
        <li
          className="listItem"
        >
          <a
            href="/letter/v"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            v
          </a>
        </li>
        <li
          className="listItem"
        >
          <a
            href="/letter/y"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            y
          </a>
        </li>
        <li
          className="listItem"
        >
          <a
            href="/letter/y2"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            ý
          </a>
        </li>
        <li
          className="listItem"
        >
          <a
            href="/letter/th"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            þ
          </a>
        </li>
        <li
          className="listItem"
        >
          <a
            href="/letter/ae"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            æ
          </a>
        </li>
        <li
          className="listItem"
        >
          <a
            href="/letter/oe3"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            œ
          </a>
        </li>
        <li
          className="listItem"
        >
          <a
            href="/letter/oe"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            ö
          </a>
        </li>
      </ul>
      <form
        onSubmit={[Function]}
      >
        <input
          name="search"
          onChange={[Function]}
          type="text"
          value=""
        />
        <button
          type="submit"
        >
          Search
        </button>
      </form>
    </nav>
    <div>
      <a
        href="/"
        onClick={[Function]}
        onMouseEnter={[Function]}
      >
        Cleasby & Vigfusson Dictionary
      </a>
      <a
        href="/letter/S"
        onClick={[Function]}
        onMouseEnter={[Function]}
      >
        Letter S
      </a>
      <a
        href="/word/SKILJA"
        onClick={[Function]}
        onMouseEnter={[Function]}
      >
        Skilja
      </a>
      <script
        dangerouslySetInnerHTML={
          Object {
            "__html": "{\\"@context\\":\\"https://schema.org/\\",\\"@type\\":\\"BreadcrumbList\\",\\"itemListElement\\":[{\\"@type\\":\\"ListItem\\",\\"position\\":1,\\"name\\":\\"Cleasby & Vigfusson Dictionary\\",\\"item\\":\\"undefined/\\"},{\\"@type\\":\\"ListItem\\",\\"position\\":2,\\"name\\":\\"Letter S\\",\\"item\\":\\"undefined/letter/S\\"},{\\"@type\\":\\"ListItem\\",\\"position\\":3,\\"name\\":\\"Skilja\\",\\"item\\":\\"undefined/word/SKILJA\\"}]}",
          }
        }
        type="application/ld+json"
      />
    </div>
  </main>,
  <footer>
    <hr />
    <section>
      <h2>
        About
      </h2>
      <p>
        Based on Cleasby & Vigfusson Dictionary Old Norse dictionary.
      </p>
      <p>
        <em>
          Icelandic-English
        </em>
         dictionary was started by Richard Cleasby and finished by Gudbrand Vigfusson.
      </p>
      <p>
        It was published in 1874, which leads to there being many public domain versions of the book available.
      </p>
    </section>
    <nav>
      <p>
        Dictionary project
      </p>
      <ul>
        <li>
          <a
            href="https://github.com/stscoundrel/cleasby-vigfusson-next"
            rel="noopener noreferrer"
            target="_blank"
          >
            Source code
          </a>
        </li>
        <li>
          <a
            href="https://github.com/stscoundrel/cleasby-vigfusson-dictionary"
            rel="noopener noreferrer"
            target="_blank"
          >
            Data source
          </a>
        </li>
      </ul>
    </nav>
    <nav>
      <p>
        Quick links
      </p>
      <ul>
        <li>
          <a
            href="/letter/a"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            a
          </a>
        </li>
        <li>
          <a
            href="/letter/a2"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            á
          </a>
        </li>
        <li>
          <a
            href="/letter/b"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            b
          </a>
        </li>
        <li>
          <a
            href="/letter/d"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            d
          </a>
        </li>
        <li>
          <a
            href="/letter/e"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            e
          </a>
        </li>
        <li>
          <a
            href="/letter/e2"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            é
          </a>
        </li>
        <li>
          <a
            href="/letter/f"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            f
          </a>
        </li>
        <li>
          <a
            href="/letter/g"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            g
          </a>
        </li>
        <li>
          <a
            href="/letter/h"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            h
          </a>
        </li>
        <li>
          <a
            href="/letter/i"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            i
          </a>
        </li>
        <li>
          <a
            href="/letter/i2"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            í
          </a>
        </li>
        <li>
          <a
            href="/letter/j"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            j
          </a>
        </li>
        <li>
          <a
            href="/letter/k"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            k
          </a>
        </li>
        <li>
          <a
            href="/letter/l"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            l
          </a>
        </li>
        <li>
          <a
            href="/letter/m"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            m
          </a>
        </li>
        <li>
          <a
            href="/letter/n"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            n
          </a>
        </li>
        <li>
          <a
            href="/letter/o"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            o
          </a>
        </li>
        <li>
          <a
            href="/letter/o2"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            ó
          </a>
        </li>
        <li>
          <a
            href="/letter/p"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            p
          </a>
        </li>
        <li>
          <a
            href="/letter/r"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            r
          </a>
        </li>
        <li>
          <a
            href="/letter/s"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            s
          </a>
        </li>
        <li>
          <a
            href="/letter/t"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            t
          </a>
        </li>
        <li>
          <a
            href="/letter/u"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            u
          </a>
        </li>
        <li>
          <a
            href="/letter/u2"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            ú
          </a>
        </li>
        <li>
          <a
            href="/letter/v"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            v
          </a>
        </li>
        <li>
          <a
            href="/letter/y"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            y
          </a>
        </li>
        <li>
          <a
            href="/letter/y2"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            ý
          </a>
        </li>
        <li>
          <a
            href="/letter/th"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            þ
          </a>
        </li>
        <li>
          <a
            href="/letter/ae"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            æ
          </a>
        </li>
        <li>
          <a
            href="/letter/oe3"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            œ
          </a>
        </li>
        <li>
          <a
            href="/letter/oe"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            ö
          </a>
        </li>
      </ul>
    </nav>
    <small>
      Copyright © 2020 Sampo Silvennoinen / StScoundrel
    </small>
  </footer>,
]
`;
